### ISSUE TITLE
[SCM-763] Password masking for svnexe does not handle all cases

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-scm#970]
**[Tobias Kalmes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tkalmes)** opened **[SCM-763](https://issues.apache.org/jira/browse/SCM-763?redirect=false)** and commented

Passwords are not masked in the log output on Linux machines. The masking works as intended on Windows machines. On linux machines tho the password is printed in clear text. This seems to be a problem due to the additional single quotes that are added around the parameters on linux machines.

---

**Affects:** 1.9

**Issue Links:**
- [MSCMPUB-37](https://issues.apache.org/jira/browse/MSCMPUB-37) Upgrade to SCM 1.11.2
 (_**"Is contained by"**_)

4 votes, 4 watchers

**Comment by jira-importer** [comment]
**[Weston Bustraan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wbustraan)** commented

This also occurs on Macs.

The culprit is actually `org.apache.maven.scm.provider.svn.svnexe.command.SvnCommandLineUtils.cryptPassword(Commandline)`

It has a rather... naïve, to be polite, implementation of the password masking. It only works if there is _exactly_ one space after `--password`. Any other condition and the password is not masked.

So, if the command line string is this:

```
svn --username myusername --password swordfish --no-auth-cache --non-interactive --trust-server-cert info
```

... the output is:

```
svn --username myusername --password '*****' --no-auth-cache --non-interactive --trust-server-cert info
```

However, it appears that, at some point, a change was made elsewhere that wraps everything in quotes on *nix OSes:

```
'svn' '--username' 'myusername' '--password' 'swordfish' '--no-auth-cache' '--non-interactive' '--trust-server-cert' 'info'
```

Now, since `--password` is followed immediately by a single quote, instead of a single space, the mask is inserted but does not replace the actual password:

```
'svn' '--username' 'myusername' '--password''*****' 'swordfish' '--no-auth-cache' '--non-interactive' '--trust-server-cert' 'info'
```

Here is an improved version of `cryptPassword` using a regex in order to handle more diverse input:

```
    public static String cryptPassword( Commandline cl )
    {
        String clString = cl.toString();
        final String mask = "'******'";

        final Matcher matcher = Pattern.compile("(--password\\S*?\\s+)('[^']+?'|\"[^\"]+?\"|\\S+)")
                                       .matcher(clString);

        final StringBuffer replaced = new StringBuffer();
        while (matcher.find()) {
            final String argPrefix = matcher.group(1);
            matcher.appendReplacement(replaced, argPrefix + mask);
        }
        matcher.appendTail(replaced);

        return replaced.toString();
    }
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Weston Bustraan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wbustraan), your improved version does not work:

```java
public static void main(String[] args) {

	Commandline cl = new Commandline(new CmdShell());
	cl.setExecutable("svn");
	cl.createArg().setValue("--password");
	cl.createArg().setValue("mumu");
	System.out.println(cl);
	String clString = cl.toString();
	final String mask = "'******'";

	final Matcher matcher = Pattern.compile("(--password\\S*?\\s+)('[^']+?'|\"[^\"]+?\"|\\S+)")
			.matcher(clString);

	final StringBuffer replaced = new StringBuffer();
	while (matcher.find()) {
		final String argPrefix = matcher.group(1);
		matcher.appendReplacement(replaced, argPrefix + mask);
	}
	matcher.appendTail(replaced);

	System.out.println(replaced.toString());

}
```

```
cmd.exe /X /C "svn --password mumu"
cmd.exe /X /C "svn --password '******'
```

It also changes the quoting style on Windows if the password contains a space.

**Comment by jira-importer** [comment]
**[Weston Bustraan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wbustraan)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), you are correct, the regex does treat the ending double quote as part of the password in your example; you would probably need a more complex routine or regex to perfectly handle that. However, the string produced by the password masking isn't intended to actually be executed by the OS; it is just printed on stdout for the user to see. The goal is to prevent the password from being printed to the log output and I believe that it still accomplishes that goal better than the original implementation.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

For the record, it is also broken on Windows too:

```
PS D:\Entwicklung\Projekte\scm-svn-test-at-sign> mvn scm:list "-Dpassword=mumu ab"
[INFO] Scanning for projects...
[INFO]
[INFO] ------------< net.sf.michael-o.dirctxsrc:dircontextsource >-------------
[INFO] Building dircontextsource 2.1.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-scm-plugin:1.9.6-SNAPSHOT:list (default-cli) @ dircontextsource ---
[INFO] Executing: cmd.exe /X /C "svn --password ***** ab" --no-auth-cache --non-interactive list --recursive file:///D:/Entwicklung/svn-repos/scm-svn-test-at-sign/branches/toll2/."
[INFO] Working directory: C:\Users\mosipov\AppData\Local\Temp
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

The following has been tried in Konsole on Fedora 27:

```
mosipov@localhost SCM-763]$ mvn scm:list  -Dpassword=toll
[INFO] Scanning for projects...
[INFO]
[INFO] ------------< net.sf.michael-o.dirctxsrc:dircontextsource >-------------
[INFO] Building dircontextsource 2.1.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-scm-plugin:1.9.6-SNAPSHOT:list (default-cli) @ dircontextsource ---
[INFO] Executing: /bin/sh -c cd /tmp && svn --password '*****' --no-auth-cache --non-interactive list --recursive file:///D:/Entwicklung/svn-repos/scm-svn-test-at-sign/branches/toll2/.
[ERROR] Provider message:
[ERROR] The svn command failed.
[ERROR] Command output:
[ERROR] svn: E170013: Unable to connect to a repository at URL 'file:///D:/Entwicklung/svn-repos/scm-svn-test-at-sign/branches/toll2'
svn: E180001: Unable to open repository 'file:///D:/Entwicklung/svn-repos/scm-svn-test-at-sign/branches/toll2'
```

as well as with space

```
mosipov@localhost SCM-763]$ mvn scm:list  -Dpassword=toll
[INFO] Scanning for projects...
[INFO]
[INFO] ------------< net.sf.michael-o.dirctxsrc:dircontextsource >-------------
[INFO] Building dircontextsource 2.1.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-scm-plugin:1.9.6-SNAPSHOT:list (default-cli) @ dircontextsource ---
[INFO] Executing: /bin/sh -c cd /tmp && svn --password '*****' --no-auth-cache --non-interactive list --recursive file:///D:/Entwicklung/svn-repos/scm-svn-test-at-sign/branches/toll2/.
[ERROR] Provider message:
[ERROR] The svn command failed.
[ERROR] Command output:
[ERROR] svn: E170013: Unable to connect to a repository at URL 'file:///D:/Entwicklung/svn-repos/scm-svn-test-at-sign/branches/toll2'
svn: E180001: Unable to open repository 'file:///D:/Entwicklung/svn-repos/scm-svn-test-at-sign/branches/toll2'
```

Yes, there are shortcoming, but I cannot reproduce the depicted case. Can someone provide a test sample?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I have pushed now the branch SCM-763 which should solve the problem. It does for me on Windows and Fedora 27 with the Maven SCM Plugin. Snapshots deployed, please test within a week.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [60d98849047f11daec78e2fb14938c12776e68c3](https://gitbox.apache.org/repos/asf?p=maven-scm.git;a=commit;h=60d98849047f11daec78e2fb14938c12776e68c3).
